### PR TITLE
current weather info

### PR DESCRIPTION
**Pull request by AlexeyAnurin**

### CODE REVIEW

File: app/src/main/java/com/griddynamics/weatherappinternship/di/AppModule.kt [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
 import com.griddynamics.weatherappinternship.BaseViewModel
+import com.griddynamics.weatherappinternship.di.scope.PerApplication
 import com.griddynamics.weatherappinternship.presentation.mainScreen.MainScreenViewModel
 import dagger.Binds
 import dagger.Module
 
 @Module
 abstract class AppModule {
 
+    @PerApplication
```
Comment (sin-al-vl) [inline app/src/main/java/com/griddynamics/weatherappinternship/di/AppModule.kt:12]: Justify this annotation here, please. Why do you think, MainScreenViewModel should be scopped to the App level?
One more thing: you've missed scope annotation in `RepositoryModule` for `WeatherRepository`.

File: app/src/main/java/com/griddynamics/weatherappinternship/data/network/interceptors/ApiKeyInterceptor.kt [hunk 2]
```diff
@@ -15,6 +17,7 @@ internal class ApiKeyInterceptor @Inject constructor() : Interceptor {
         val url = chainRequest.url
                 .newBuilder()
                 .addQueryParameter(API_KEY, APP_ID)
+                .addQueryParameter(UNITS, CELSIUS)
```
Comment (sin-al-vl) [inline app/src/main/java/com/griddynamics/weatherappinternship/data/network/interceptors/ApiKeyInterceptor.kt:20]: Class called `ApiKeyInterceptor`, but now we work with metric here. Don't you find this confusing?

File: app/src/main/java/com/griddynamics/weatherappinternship/domain/model/WeatherModel.kt [hunk 3]
```diff
@@ -5,7 +5,94 @@ import com.squareup.moshi.JsonClass
 
 @JsonClass(generateAdapter = true)
```
Comment (sin-al-vl) [inline app/src/main/java/com/griddynamics/weatherappinternship/domain/model/WeatherModel.kt:6]: Domain layer shouldn't know anything about any frameworks (Moshi in this case).

File: app/src/main/java/com/griddynamics/weatherappinternship/presentation/MainActivity.kt [hunk 4]
```diff
@@ -2,16 +2,40 @@ package com.griddynamics.weatherappinternship.presentation
...
+
     override fun onCreate(savedInstanceState: Bundle?) {
+        (application as BaseApplication).appComp().inject(this)
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
-        supportFragmentManager.beginTransaction()
-                .add(R.id.fragment_container_view, MainScreenFragment())
-                .commit()
+
+        val sectionsPagerAdapter = SectionsPagerAdapter(this, supportFragmentManager)
```
Comment (sin-al-vl) [inline app/src/main/java/com/griddynamics/weatherappinternship/presentation/MainActivity.kt:24]: It's cheating) Current weather info ticket doesn't imply any pager adapter implementation. You've mixed here another task that you even don't have on the board yet.

File: app/src/main/java/com/griddynamics/weatherappinternship/presentation/MainActivity.kt [hunk 5]
```diff
@@ -2,16 +2,40 @@ package com.griddynamics.weatherappinternship.presentation
...
+        val viewPager: ViewPager = findViewById(R.id.view_pager)
+        viewPager.adapter = sectionsPagerAdapter
+        val tabs: TabLayout = findViewById(R.id.tabs)
+        tabs.setupWithViewPager(viewPager)
+
+        sv_city_search.setOnQueryTextListener(object : android.widget.SearchView.OnQueryTextListener {
```
Comment (sin-al-vl) [inline app/src/main/java/com/griddynamics/weatherappinternship/presentation/MainActivity.kt:30]: You should mention in the MR description, that `Place search by name` is also implemented here. Now it looks confusing to me.
By the way, your commit message is also quite confusing and misleading.

File: app/src/main/java/com/griddynamics/weatherappinternship/presentation/SectionsPagerAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,30 @@
+package com.griddynamics.weatherappinternship.presentation
+
+import android.content.Context
+import androidx.fragment.app.Fragment
+import androidx.fragment.app.FragmentManager
+import androidx.fragment.app.FragmentPagerAdapter
+import com.griddynamics.weatherappinternship.R
+import com.griddynamics.weatherappinternship.presentation.mainScreen.MainScreenFragment
+
+private val TAB_TITLES = arrayOf(
```
Comment (sin-al-vl) [inline app/src/main/java/com/griddynamics/weatherappinternship/presentation/SectionsPagerAdapter.kt:10]: Check previous comments.

File: app/src/main/java/com/griddynamics/weatherappinternship/presentation/mainScreen/MainScreenFragment.kt [hunk 7]
```diff
@@ -33,10 +34,31 @@ class MainScreenFragment : BaseFragment() {
...
-        mainScreenViewModel.getWeather("New York")
-
         mainScreenViewModel.weatherEvent.observe(viewLifecycleOwner, Observer {
-            tv_mainFragment.text = it.name
+            tv_current_temp.text = it.main.temperature.toInt().toString()
+            tv_feels_like.text = it.main.feelsLike.toInt().toString()
+            tv_description_weather.text = it.weather[0].weatherDescription
+            Glide.with(this)
+                    .load("https://openweathermap.org/img/wn/" +
+                            "${it.weather[0].weatherIcon}@2x.png")
```
Comment (sin-al-vl) [inline app/src/main/java/com/griddynamics/weatherappinternship/presentation/mainScreen/MainScreenFragment.kt:43]: I believe, this url should be formed somewhere on the repo lvl.

File: app/src/main/res/layout/activity_main.xml [hunk 8]
```diff
@@ -1,15 +1,51 @@
...
-        android:id="@+id/fragment_container_view"
+    <androidx.constraintlayout.widget.ConstraintLayout
         android:layout_width="match_parent"
-        android:layout_height="match_parent" />
+        android:layout_height="match_parent">
 
-</androidx.constraintlayout.widget.ConstraintLayout>
\ No newline at end of file
+        <com.google.android.material.appbar.AppBarLayout
+            android:id="@+id/ab_appBar"
```
Comment (sin-al-vl) [inline app/src/main/res/layout/activity_main.xml:14]: What `ab` prefix means? It's not obvious foe me

File: app/src/main/res/layout/activity_main.xml [hunk 9]
```diff
@@ -1,15 +1,51 @@
...
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:theme="@style/Theme.WeatherAppInternship.AppBarOverlay"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toTopOf="parent">
+
+            <SearchView
+                android:id="@+id/sv_city_search"
```
Comment (sin-al-vl) [inline app/src/main/res/layout/activity_main.xml:22]: If `sv` means `search view`, I'd say that suffix `search` is redundant

File: app/src/main/res/layout/activity_main.xml [hunk 10]
```diff
@@ -1,15 +1,51 @@
...
+                android:layout_width="match_parent"
+                android:layout_height="50dp"
+                android:background="@color/gray1"
+                app:layout_constraintStart_toStartOf="parent"
+                app:layout_constraintTop_toBottomOf="@id/ab_appBar" />
+
+            <com.google.android.material.tabs.TabLayout
+                android:id="@+id/tabs"
```
Comment (sin-al-vl) [inline app/src/main/res/layout/activity_main.xml:30]: AppBarLayout has one principle of the id name forming. TabLayout has another principle. It's different UI elements, but they are both layouts. Please, be consistent in naming policy.

File: app/src/main/res/layout/activity_main.xml [hunk 11]
```diff
@@ -1,15 +1,51 @@
...
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                app:tabBackground="@color/background_tabLayout"
```
Comment (sin-al-vl) [inline app/src/main/res/layout/activity_main.xml:33]: Camel case + snake case in the same ref name. This is unacceptable

File: app/src/main/res/layout/fragment_mainscreen.xml [hunk 12]
```diff
@@ -1,18 +1,85 @@
...
-    <TextView
-        android:id="@+id/tv_mainFragment"
-        android:layout_width="match_parent"
+    <TextClock
+        android:id="@+id/tv_date_time"
+        android:layout_width="wrap_content"
         android:layout_height="wrap_content"
-        android:text="MAINFRAMENT"
+        android:layout_marginStart="16dp"
+        android:layout_marginTop="16dp"
```
Comment (sin-al-vl) [inline app/src/main/res/layout/fragment_mainscreen.xml:14]: If you see frequently used indentations, please, move their values to `dimens`

File: app/src/main/res/layout/fragment_mainscreen.xml [hunk 13]
```diff
@@ -1,18 +1,85 @@
...
+        android:format12Hour="d MMM, HH:mm"
+        android:format24Hour="EEEE, MMMM dd, yyyy h:mmaa"
+        android:text="Date &amp; Time"
```
Comment (sin-al-vl) [inline app/src/main/res/layout/fragment_mainscreen.xml:17]: string res

File: app/src/main/res/values/strings.xml [hunk 14]
```diff
@@ -1,3 +1,11 @@
 <resources>
     <string name="app_name">WeatherAppInternship</string>
+    <string name="tab1_text">TAB_1</string>
+    <string name="tab_text_1">TAB_1</string>
+    <string name="tab_text_2">TAB_2</string>
+    <string name="tab_text_3">TAB_3</string>
+    <string name="feels_like">Ощущается как</string>
```
Comment (sin-al-vl) [inline app/src/main/res/values/strings.xml:7]: Quite unexpectedly. What Russian text does among English string values?

**Review by sin-al-vl — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/griddynamics/weatherappinternship/data/network/interceptors/ApiKeyInterceptor.kt (hunk 2 above)
Comment (AlexeyAnurin) [reply app/src/main/java/com/griddynamics/weatherappinternship/data/network/interceptors/ApiKeyInterceptor.kt:20]: Separate UnitsInterceptor created. fixed.

File: app/src/main/java/com/griddynamics/weatherappinternship/domain/model/WeatherModel.kt (hunk 3 above)
Comment (AlexeyAnurin) [reply app/src/main/java/com/griddynamics/weatherappinternship/domain/model/WeatherModel.kt:6]: Dto class with moshi annotation created. Values mapped to domain model.

File: app/src/main/java/com/griddynamics/weatherappinternship/presentation/mainScreen/MainScreenFragment.kt (hunk 7 above)
Comment (AlexeyAnurin) [reply app/src/main/java/com/griddynamics/weatherappinternship/presentation/mainScreen/MainScreenFragment.kt:43]: I didnt find another place for url building, but buit it with constants.

File: app/src/main/res/layout/fragment_mainscreen.xml (hunk 13 above)
Comment (AlexeyAnurin) [reply app/src/main/res/layout/fragment_mainscreen.xml:17]: fixed

File: app/src/main/res/layout/fragment_mainscreen.xml (hunk 12 above)
Comment (AlexeyAnurin) [reply app/src/main/res/layout/fragment_mainscreen.xml:14]: fixed

File: app/src/main/res/values/strings.xml (hunk 14 above)
Comment (AlexeyAnurin) [reply app/src/main/res/values/strings.xml:7]: fixed

File: app/src/main/res/layout/activity_main.xml (hunk 9 above)
Comment (AlexeyAnurin) [reply app/src/main/res/layout/activity_main.xml:22]: fixed

File: app/src/main/res/layout/activity_main.xml (hunk 11 above)
Comment (AlexeyAnurin) [reply app/src/main/res/layout/activity_main.xml:33]: fixed

File: app/src/main/res/layout/activity_main.xml (hunk 10 above)
Comment (AlexeyAnurin) [reply app/src/main/res/layout/activity_main.xml:30]: fixed

File: app/src/main/res/layout/activity_main.xml (hunk 9 above)
Comment (AlexeyAnurin) [reply app/src/main/res/layout/activity_main.xml:22]: I would name in "city_search_view"

File: app/src/main/java/com/griddynamics/weatherappinternship/presentation/mainScreen/MainScreenFragment.kt (hunk 7 above)
Comment (AlexeyAnurin) [reply app/src/main/java/com/griddynamics/weatherappinternship/presentation/mainScreen/MainScreenFragment.kt:43]: I would name in "appbar_layout"

File: app/src/main/res/layout/activity_main.xml (hunk 8 above)
Comment (AlexeyAnurin) [reply app/src/main/res/layout/activity_main.xml:14]: I would name in "appbar_layout"

File: app/src/main/java/com/griddynamics/weatherappinternship/di/AppModule.kt (hunk 1 above)
Comment (AlexeyAnurin) [reply app/src/main/java/com/griddynamics/weatherappinternship/di/AppModule.kt:12]: I have added MainActivityComponent as a SubComponent for scope @PerActivity to scope there ViewModel.

File: app/src/main/java/com/griddynamics/weatherappinternship/presentation/MainActivity.kt (hunk 4 above)
Comment (AlexeyAnurin) [reply app/src/main/java/com/griddynamics/weatherappinternship/presentation/MainActivity.kt:24]: Deleted from this PR

File: app/src/main/java/com/griddynamics/weatherappinternship/presentation/MainActivity.kt (hunk 5 above)
Comment (AlexeyAnurin) [reply app/src/main/java/com/griddynamics/weatherappinternship/presentation/MainActivity.kt:30]: Deleted from this PR

File: app/src/main/java/com/griddynamics/weatherappinternship/presentation/SectionsPagerAdapter.kt (hunk 6 above)
Comment (AlexeyAnurin) [reply app/src/main/java/com/griddynamics/weatherappinternship/presentation/SectionsPagerAdapter.kt:10]: Deleted from this PR

File: app/src/main/java/com/griddynamics/weatherappinternship/data/network/model/WeatherModelDto.kt [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+    @Json(name = "id")
+    val id: Int,
+    @Json(name = "name")
+    val name: String,
+    @Json(name = "cod")
+    val cod: Int
+)
+
+@JsonClass(generateAdapter = true)
+data class Coordinates(
```
Comment (sin-al-vl) [inline app/src/main/java/com/griddynamics/weatherappinternship/data/network/model/WeatherModelDto.kt:38]: Naming consistency. If this is a network DTO, class name should ends with Dto as well

File: app/src/main/java/com/griddynamics/weatherappinternship/data/network/model/WeatherModelDto.kt [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+    val weather = weather.first()
+    val temperatureParams = temperatureParams
+    return WeatherModel(
+        weatherId = weather.weatherId,
+        weatherDescription = weather.weatherDescription,
+        weatherIcon = weather.weatherIcon,
+        temperature = temperatureParams.temperature.toInt().toString(),
+        feelsLike = temperatureParams.feelsLike.toInt().toString()
+    )
+}
```
Comment (sin-al-vl) [inline app/src/main/java/com/griddynamics/weatherappinternship/data/network/model/WeatherModelDto.kt:111]: As I said earlier, this function should be moved out to the separate file in `network.mapper` package.

File: app/src/main/java/com/griddynamics/weatherappinternship/data/network/model/WeatherModelDto.kt [hunk 17]
```diff
@@ -0,0 +1,111 @@
+package com.griddynamics.weatherappinternship.data.network.model
+
+import com.griddynamics.weatherappinternship.domain.model.WeatherModel
+import com.squareup.moshi.Json
+import com.squareup.moshi.JsonClass
+
+@JsonClass(generateAdapter = true)
+data class WeatherModelDto(
```
Comment (sin-al-vl) [inline app/src/main/java/com/griddynamics/weatherappinternship/data/network/model/WeatherModelDto.kt:8]: `internal`. The same for all classes in this package

File: app/src/main/java/com/griddynamics/weatherappinternship/data/repository/WeatherRepositoryImpl.kt [hunk 18]
```diff
@@ -4,16 +4,20 @@ import com.griddynamics.weatherappinternship.domain.ResultObject
...
 @PerApplication
-internal class WeatherRepositoryImpl @Inject constructor(private val weatherApi: WeatherApi): WeatherRepository {
+internal class WeatherRepositoryImpl @Inject constructor(
+    private val weatherApi: WeatherApi
+) : WeatherRepository {
 
     override suspend fun getWeather(cityName: String): ResultObject<WeatherModel> {
         return try {
             val result = weatherApi.getWeather(cityName)
+                .toDomain()
```
Comment (sin-al-vl) [inline app/src/main/java/com/griddynamics/weatherappinternship/data/repository/WeatherRepositoryImpl.kt:20]: That's exact case, where `.toDomain` should stay on the previous line

File: app/src/main/java/com/griddynamics/weatherappinternship/presentation/mainScreen/MainScreenFragment.kt [hunk 19]
```diff
@@ -6,37 +6,65 @@ import android.view.LayoutInflater
...
+import com.bumptech.glide.Glide
 import com.griddynamics.weatherappinternship.BaseFragment
 import com.griddynamics.weatherappinternship.R
-import kotlinx.android.synthetic.main.fragment_mainscreen.*
+import com.griddynamics.weatherappinternship.databinding.FragmentMainscreenBinding
+import com.griddynamics.weatherappinternship.presentation.MainActivity
 import javax.inject.Inject
 
+private const val BASE_WEATHER_ICON_URL = "https://openweathermap.org/img/wn/"
+private const val BASE_WEATHER_ICON_PROPERTIES = "@2x.png"
```
Comment (sin-al-vl) [inline app/src/main/java/com/griddynamics/weatherappinternship/presentation/mainScreen/MainScreenFragment.kt:17]: What would you do, when there will be need to display such icon on another fragment? This should be moved to mapper. And `BASE_WEATHER_ICON_URL + it.weatherIcon + BASE_WEATHER_ICON_PROPERTIES` transformation should be there as well.

File: app/src/main/res/values/strings.xml [hunk 20]
```diff
@@ -1,3 +1,15 @@
 <resources>
     <string name="app_name">WeatherAppInternship</string>
+    <string name="tab1_text">TAB_1</string>
+    <string name="tab_text_1">TAB_1</string>
+    <string name="tab_text_2">TAB_2</string>
+    <string name="tab_text_3">TAB_3</string>
```
Comment (sin-al-vl) [inline app/src/main/res/values/strings.xml:6]: Why do we need this strings?

**Review by sin-al-vl — CHANGES_REQUESTED** [review]

File: app/src/main/res/values/strings.xml [hunk 21]
```diff
@@ -1,3 +1,15 @@
...
+    <string name="feels_like">feels like</string>
+    <string name="temp_main_screen">00</string>
+    <string name="degree_sign">℃</string>
+    <string name="weather_type">weather type</string>
+    <string name="error_message">Error message</string>
+    <string name="date_amp_time"><![CDATA[Date & Time]]></string>
+    <string name="format12Hours">d MMM, HH:mm</string>
+    <string name="format24Hour">EEEE, MMMM dd, yyyy h:mmaa</string>
```
Comment (sin-al-vl) [inline app/src/main/res/values/strings.xml:14]: Such patterns should be marked with `translatable="false"`. And it's better to keep them in separate string res file with the name `donottranslate.xml`, for example

**Review by sin-al-vl — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/griddynamics/weatherappinternship/data/network/model/WeatherModelDto.kt (hunk 15 above)
Comment (AlexeyAnurin) [reply app/src/main/java/com/griddynamics/weatherappinternship/data/network/model/WeatherModelDto.kt:38]: fixed

File: app/src/main/java/com/griddynamics/weatherappinternship/data/repository/WeatherRepositoryImpl.kt (hunk 18 above)
Comment (AlexeyAnurin) [reply app/src/main/java/com/griddynamics/weatherappinternship/data/repository/WeatherRepositoryImpl.kt:20]: fixed

File: app/src/main/res/values/strings.xml (hunk 20 above)
Comment (AlexeyAnurin) [reply app/src/main/res/values/strings.xml:6]: Fixed.

File: app/src/main/java/com/griddynamics/weatherappinternship/data/network/model/WeatherModelDto.kt (hunk 17 above)
Comment (AlexeyAnurin) [reply app/src/main/java/com/griddynamics/weatherappinternship/data/network/model/WeatherModelDto.kt:8]: When I mark classes as internal, an error comes
image

File: app/src/main/java/com/griddynamics/weatherappinternship/data/network/model/WeatherModelDto.kt (hunk 16 above)
Comment (AlexeyAnurin) [reply app/src/main/java/com/griddynamics/weatherappinternship/data/network/model/WeatherModelDto.kt:111]: fun WeatherModelDto.toDomain() is transferred to file WeatherToDomainMapper in network.mapper package.

File: app/src/main/java/com/griddynamics/weatherappinternship/presentation/mainScreen/MainScreenFragment.kt (hunk 19 above)
Comment (AlexeyAnurin) [reply app/src/main/java/com/griddynamics/weatherappinternship/presentation/mainScreen/MainScreenFragment.kt:17]: Extension function for building icon url is added to mapper.

File: app/src/main/res/values/strings.xml (hunk 21 above)
Comment (AlexeyAnurin) [reply app/src/main/res/values/strings.xml:14]: fixed

File: app/src/main/java/com/griddynamics/weatherappinternship/data/network/model/WeatherModelDto.kt (hunk 17 above)
Comment (sin-al-vl) [reply app/src/main/java/com/griddynamics/weatherappinternship/data/network/model/WeatherModelDto.kt:8]: WeatherApi should be internal as well. And so, and so...

Comment (AlexeyAnurin) [reply app/src/main/java/com/griddynamics/weatherappinternship/data/network/model/WeatherModelDto.kt]: fixed

File: app/src/main/java/com/griddynamics/weatherappinternship/presentation/mainScreen/MainScreenFragment.kt [hunk 22]
```diff
@@ -6,37 +6,64 @@ import android.view.LayoutInflater
...
+import com.bumptech.glide.Glide
 import com.griddynamics.weatherappinternship.BaseFragment
 import com.griddynamics.weatherappinternship.R
-import kotlinx.android.synthetic.main.fragment_mainscreen.*
+import com.griddynamics.weatherappinternship.databinding.FragmentMainscreenBinding
+import com.griddynamics.weatherappinternship.presentation.MainActivity
 import javax.inject.Inject
 
+/*private const val BASE_WEATHER_ICON_URL = "https://openweathermap.org/img/wn/"
+private const val BASE_WEATHER_ICON_PROPERTIES = "@2x.png"*/
```
Comment (sin-al-vl) [inline app/src/main/java/com/griddynamics/weatherappinternship/presentation/mainScreen/MainScreenFragment.kt:17]: no way

File: app/src/main/java/com/griddynamics/weatherappinternship/domain/model/WeatherModel.kt [hunk 23]
```diff
@@ -1,11 +1,9 @@
...
-
-@JsonClass(generateAdapter = true)
 data class WeatherModel(
-        @Json(name = "name")
-        val name: String
-)
-
+    val weatherId: Int,
+    val weatherDescription: String,
+    val weatherIcon: String,
```
Comment (sin-al-vl) [inline app/src/main/java/com/griddynamics/weatherappinternship/domain/model/WeatherModel.kt:6]: weatherIconUrl

**Review by sin-al-vl — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/griddynamics/weatherappinternship/presentation/mainScreen/MainScreenFragment.kt (hunk 22 above)
Comment (AlexeyAnurin) [reply app/src/main/java/com/griddynamics/weatherappinternship/presentation/mainScreen/MainScreenFragment.kt:17]: fixed

File: app/src/main/java/com/griddynamics/weatherappinternship/domain/model/WeatherModel.kt (hunk 23 above)
Comment (AlexeyAnurin) [reply app/src/main/java/com/griddynamics/weatherappinternship/domain/model/WeatherModel.kt:6]: fixed

**Pull request merged**